src/Microsoft.VisualStudio.Editor.Razor/DefaultTagHelperCompletionService.cs

```diff
@@ -194,7 +194,7 @@ public override ElementCompletionResult GetElementCompletions(ElementCompletionC
                     {
                         // If there is an unknown HTML schema tag that doesn't exist in the current completion we should add it. This happens for
                         // TagHelpers that target non-schema oriented tags.
-                        addRuleCompletions = true;
+                        addRuleCompletions = TagHelperMatchingConventions.SatisfiesParentTag(completionContext.ContainingTagName, rule);
```

Comment:
FYI: This is the same condition we initially run to filter the taghelpers https://github.com/aspnet/Razor/blob/release/2.1/src/Microsoft.VisualStudio.Editor.Razor/DefaultTagHelperFactsService.cs#L155

Reply:
The `DefaultTagHelperFactsService` applies this at a broader scale (it doesn't have all of the conditions above). You need to filter every rule from the descriptor by this `SatisfiesParentTag`. Just as an example if you add `[OutputElementHint("span")]` to the `ChildTagHelper` in the originally reported issue it'll ignore the parent restriction again.